lib/auth/auth_with_roles.go

```diff
@@ -7213,3 +7251,77 @@ func verbsToReplaceResourceWithOrigin(stored types.ResourceWithOrigin) []string
 	}
 	return verbs
 }
+
+// checkOktaAccess gates access to update operations on user records based
+// on the origin label on the supplied user record.
+//
+// # See checkOktaUserAccess() for the actual access rules
+//
+// TODO(tcsc): Delete in 16.0.0 when user management is removed from `ServerWithRoles`
+func checkOktaAccess(ctx context.Context, authzCtx *authz.Context, users services.UsersService, existingUsername string, verb string) error {
+	existingUser, err := users.GetUser(ctx, existingUsername, false)
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+
+	return checkOktaUserAccess(authzCtx, existingUser, verb)
+}
+
+// checkOktaUserAccess gates access to update operations on user records based
+// on the origin labels of the existing and new user records.
+//
+// A nil `existingUser` is interpreted as there being no matching existing
+// user in the cluster; if there is no user then there is no user to
+// overwrite, so access is granted
+//
+// TODO(tcsc): Delete in 16.0.0 when user management is removed from `ServerWithRoles`
+func checkOktaUserAccess(authzCtx *authz.Context, existingUser types.User, verb string) error {
+	// We base or decision to allow write access to a resource on the Origin
+	// label. If there is no existing user, then there can be no label to block
+	// access, so anyone can do anything.
+	if existingUser == nil {
+		return nil
+	}
+
+	if !authz.HasBuiltinRole(*authzCtx, string(types.RoleOkta)) {
+		// The only thing a non-okta service caller is allowed to do to an
+		// Okta-origin user is delete it
+		if (existingUser.Origin() == types.OriginOkta) && (verb != types.VerbDelete) {
+			return trace.AccessDenied("Only the okta service may update okta users")
+		}
+		return nil
+	}
+
+	// An okta-service caller only has rights over the user if they have an
+	// "Origin: Okta" label
+	if existingUser.Origin() == types.OriginOkta {
+		return nil
+	}
+
+	// If we get to here, we have exhausted all possible ways that the caller
+	// may be allowed to modify a user, so they get AccessDenied by default.
+	return trace.AccessDenied("Okta service may only %s Okta users", verb)
+}
+
+// checkOktaOrigin checks that the supplied user has an appropriate origin label
+// set. In this case "appropriate" means having the Okta origin set if and only
+// if the supplied auth context has the build-in Okta role. Context without the
+// Okta role may supply any origin value *other than* okta (including nil).
+// Returns an error if the user origin value is "inappropriate".
+//
+// TODO(tcsc): Delete in 16.0.0 when user management is removed from `ServerWithRoles`
+func checkOktaOrigin(authzCtx *authz.Context, user types.User) error {
+	isOktaService := authz.HasBuiltinRole(*authzCtx, string(types.RoleOkta))
+	hasOktaOrigin := user.Origin() == types.OriginOkta
+
+	switch {
+	case isOktaService && !hasOktaOrigin:
+		return trace.BadParameter(`Okta service must supply "okta" origin`)
```

Comment:
Nit: Can't we just force-set the origin label if it's an Okta service that's creating the user?

Reply:
Definitely possible, but I'm always a bit leery of silently correcting a caller mistake. That said, it does feel a bit like hoop jumping to force it.